src/backend/base/langflow/services/job_queue/service.py

```diff
@@ -0,0 +1,254 @@
+
+        # Initiate the new asynchronous task.
+        task = asyncio.create_task(task_coro)
+        self._queues[job_id] = (main_queue, event_manager, task)
+        logger.debug(f"New task started for job_id {job_id}")
+
+    def get_queue_data(self, job_id: str) -> tuple[asyncio.Queue, EventManager, asyncio.Task | None]:
+        """Retrieve the complete data structure associated with a job's queue.
+
+        Args:
+            job_id (str): Unique identifier for the job.
+
+        Returns:
+            tuple[asyncio.Queue, EventManager, asyncio.Task | None]:
+                A tuple containing the job's main queue, its linked event manager, and the associated task (if any).
+
+        Raises:
+            ValueError: If no queue is registered for the given job_id.
+        """
+        if job_id not in self._queues:
+            msg = f"No queue found for job_id {job_id}"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        return self._queues[job_id]
+
+    async def cleanup_job(self, job_id: str) -> None:
+        """Clean up and release resources for a specific job.
+
+        The cleanup process includes:
+          1. Verifying if the job's queue is registered.
+          2. Cancelling the running task (if active) and awaiting its termination.
+          3. Clearing all items from the job's queue.
+          4. Removing the job's entry from the internal registry.
+
+        Args:
+            job_id (str): Unique identifier for the job to be cleaned up.
+        """
+        if job_id not in self._queues:
+            logger.debug(f"No queue found for job_id {job_id} during cleanup.")
+            return
+
+        logger.info(f"Commencing cleanup for job_id {job_id}")
+        main_queue, event_manager, task = self._queues[job_id]
+
+        # Cancel the associated task if it is still running.
+        if task and not task.done():
+            logger.debug(f"Cancelling active task for job_id {job_id}")
+            task.cancel()
+            with contextlib.suppress(asyncio.CancelledError):
+                await task
+            logger.debug(f"Task cancellation complete for job_id {job_id}")
+
+        # If there is no active task or the task has completed, proceed with clearing the queue.
+        if task is None or task.done():
+            items_cleared = 0
+            while not main_queue.empty():
+                try:
+                    main_queue.get_nowait()
+                    items_cleared += 1
+                except asyncio.QueueEmpty:
+                    break
+
+            logger.debug(f"Removed {items_cleared} items from queue for job_id {job_id}")
+            # Delete the job entry from the registry.
+            del self._queues[job_id]
+            logger.info(f"Cleanup successful for job_id {job_id}: resources have been released.")
+        else:
+            logger.warning(f"Cleanup for job_id {job_id} deferred: task still running. Will retry in the next cycle.")
```

Comment:
I don't think it's possible to get here.

Reply:
Agree.